core/notification/message.go

```diff
@@ -0,0 +1,132 @@
+package notification
+
+import (
+	"time"
+
+	"github.com/google/uuid"
+)
+
+// MessageStatus determines the state of the message
+type MessageStatus string
+
+const (
+	DefaultMaxTries = 3
+
+	MessageStatusEnqueued  MessageStatus = "enqueued"
+	MessageStatusFailed    MessageStatus = "failed"
+	MessageStatusPending   MessageStatus = "pending"
+	MessageStatusPublished MessageStatus = "published"
+)
+
+// MessageOption provides ability to configure the message initialization
+type MessageOption func(*Message)
+
+// InitWithCreateTime initializes the message with custom create time
+func InitWithCreateTime(timeNow time.Time) MessageOption {
+	return func(m *Message) {
+		m.CreatedAt = timeNow
+		m.UpdatedAt = timeNow
+	}
+}
+
+// InitWithID initializes the message with some ID
+func InitWithID(id string) MessageOption {
+	return func(m *Message) {
+		m.ID = id
+	}
+}
+
+// InitWithExpiryDuration initializes the message with the specified expiry duration
+func InitWithExpiryDuration(dur time.Duration) MessageOption {
+	return func(m *Message) {
+		m.expiryDuration = dur
+	}
+}
+
+// InitWithMaxTries initializes the message with custom max tries
+func InitWithMaxTries(mt int) MessageOption {
+	return func(m *Message) {
+		m.MaxTries = mt
+	}
+}
+
+// Message is the model to be sent for a specific subscription's receiver
+type Message struct {
+	ID     string
+	Status MessageStatus
+
+	ReceiverType string
+	Configs      map[string]interface{} // the datasource to build vendor-specific configs
+	Detail       map[string]interface{} // the datasource to build vendor-specific message
+	LastError    string
+
+	MaxTries  int
+	TryCount  int
+	Retryable bool
+
+	ExpiredAt time.Time
+	CreatedAt time.Time
+	UpdatedAt time.Time
+
+	expiryDuration time.Duration
+}
+
+// Initialize initializes the message with some default value
+// or the customized value
+func (m *Message) Initialize(
+	n Notification,
+	receiverType string,
+	notificationConfigs map[string]interface{},
+	opts ...MessageOption,
+) {
+	var timeNow = time.Now()
+
+	m.ID = uuid.NewString()
+	m.Status = MessageStatusEnqueued
+
+	m.ReceiverType = receiverType
+	m.Configs = notificationConfigs
+	detail := make(map[string]interface{})
+	for k, v := range n.Labels {
+		detail[k] = v
+	}
+	for k, v := range n.Data {
+		detail[k] = v
+	}
+	m.Detail = detail
```

Comment:
to avoid key conflict, should we put each `labels` and `data` into different keys in `detail`? so there will be `detail["labels"]` for `n.Labels` and the same for data

Reply:
It is okay, for this case (transforming notification to messages), `data` would be prioritized more than `label`. The usage of `label` is only to match and passing that to message would just giving some additional details to the receiver. So it is okay to just keep `data` overwriting the `label` in this case.
Doing above would make our model unnecessary too complex.